db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/FromBaseTable.java

```diff
@@ -926,11 +1389,279 @@ public CostEstimate estimateCost(OptimizablePredicateList predList,
+                            C_NodeTypes.SELECT_NODE,
+                            resultColumnList,
+                            null,
+                            fromList,
+                            whereClause,
+                            null,
+                            null,
+                            null,
+                            getContextManager());
+       return selectNode;
+    }
+
+    // If the UIS access path table has join predicates, pull in the single-table predicates
+    // on the outer tables so they can reduce the number of rows considered in the join of each
+    // UNION branch of a UIS access path.
+    @SuppressFBWarnings(value = "RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE", justification = "intentional")
+    private boolean addPredsFromOuterRestrictionList(ProjectRestrictNode projectRestrict, ValueNode whereClause) {
+       if (projectRestrict != null) {
+           PredicateList restrictionList = null;
+           restrictionList = projectRestrict.getRestrictionList();
+           // Applying Unioned Index Scan join predicates is typically only effective
+           // if the outer table has predicates limiting the scan.
+           // Avoid costly plans with no such predicates for now, unless we can find
+           // some reason to use them in the future.
+           if (restrictionList == null || restrictionList.isEmpty())
+               return false;
+           if (restrictionList != null) {
+               CloneCRsVisitor cloneCRsVisitor = new CloneCRsVisitor();
+               cloneCRsVisitor.setInitializeSourceOfCR(true);
+               AndNode tempAnd = (AndNode)whereClause;
+   nextPred:   for(int i=0; i < restrictionList.size(); i++) {
+                   Predicate predicate = restrictionList.elementAt(i);
+                   AndNode andNodeToAdd = predicate.getAndNode();
+                   if (tempAnd == andNodeToAdd)
+                       continue nextPred;
+                   while (!tempAnd.getRightOperand().isBooleanTrue()) {
+                       tempAnd = (AndNode) tempAnd.getRightOperand();
+                       if (tempAnd == andNodeToAdd)
+                           continue nextPred;
+                   }
+                   try {
+                       andNodeToAdd = (AndNode)andNodeToAdd.accept(cloneCRsVisitor);
+                   }
+                   catch (StandardException e) {
+                       // If unable to clone all ColumnReferences, it is
+                       // unsafe to use this predicate tree.
+                       return false;
+                   }
+                   tempAnd.setRightOperand(andNodeToAdd);
+               }
+           }
+       }
+       return true;
+    }
+
+    // Build a UNION node between 2 branches in a Unioned Index Scans access path.
+    private FromTable getUnionNode(ResultSetNode leftSide, FromTable rightSide, Optimizer optimizer) throws StandardException {
```

Comment:
Function signature is strange. Why `leftSide` is `ResultSetNode` but `rightSide` is a `FromTable`?

Reply:
Changed the function signature to:

private UnionNode getUnionNode(ResultSetNode leftSide, ResultSetNode rightSide, Optimizer optimizer) throws StandardException